restli-common/src/main/java/com/linkedin/restli/internal/common/URIMaskUtil.java

```diff
@@ -115,137 +110,162 @@ else if (entry.getValue()
-              + toparse + "' while traversing");
+              + toparse.substring(index) + "' while traversing");
         }
-        toparse.delete(0, 1);
+        index++;
         state = ParseState.PARSE_FIELDS;
         break;
       case DESCEND:
-        if (toparse.indexOf(":(") != 0)
+        if (toparse.charAt(index) != ':' || toparse.charAt(index + 1) != '(')
         {
           throw new IllegalStateException("Internal Error parsing mask: unexpected parse buffer '"
-              + toparse + "' while descending");
+              + toparse.substring(index) + "' while descending");
         }
-        toparse.delete(0, 2);
+        index += 2;
         state = ParseState.PARSE_FIELDS;
         break;
       case PARSE_FIELDS:
-
-        Integer maskValue = null;
-        if (toparse.charAt(0) == '-')
+        Integer maskValue;
+        if (toparse.charAt(index) == '-')
         {
           maskValue = MaskOperation.NEGATIVE_MASK_OP.getRepresentation();
-          toparse.delete(0, 1);
+          index++;
         }
         else
         {
           maskValue = MaskOperation.POSITIVE_MASK_OP.getRepresentation();
         }
 
         int nextToken = -1;
-        StringBuilder field = new StringBuilder();
-        for (int ii = 0; ii < toparse.length(); ++ii)
+        field.setLength(0);
+        int fieldIndex = index;
+        for (; fieldIndex < toparse.length(); ++fieldIndex)
         {
-          char c = toparse.charAt(ii);
+          char c = toparse.charAt(fieldIndex);
           switch (c)
           {
           case ',':
             state = ParseState.TRAVERSE;
-            nextToken = ii;
+            nextToken = fieldIndex;
             break;
           case ':':
-            if (field.length() > 0 && (FilterConstants.START.equals(field.toString()) || FilterConstants.COUNT.equals(field.toString())))
+            if ((fieldIndex + 1) >= toparse.length())
+            {
+              throw new IllegalMaskException("Malformed mask syntax: unexpected end of buffer after ':'");
+            }
+            if ((field.length() == FilterConstants.START.length() && field.indexOf(FilterConstants.START) == 0)
+                || (field.length() == FilterConstants.COUNT.length() && field.indexOf(FilterConstants.COUNT) == 0))
             {
-              if (!Character.isDigit(toparse.charAt(ii + 1)))
+              if (!Character.isDigit(toparse.charAt(fieldIndex + 1)))
               {
                 throw new IllegalMaskException("Malformed mask syntax: unexpected range value");
               }
 
-              ii++;
+              fieldIndex++;
 
               // Aggressively consume the numerical value for the range parameter as this is a special case.
-              StringBuilder rangeValue = new StringBuilder();
-              while (ii < toparse.length())
+              int rangeValue = 0;
+              while (fieldIndex < toparse.length() && nextToken == -1)
```

Comment:
This is more like a question, `nextToken` is set when we see a comma ',',  and this while loop is condition to enter when the `nextToken` is set to `-1`. Do we design to process only the token before comma ','?

If the token list looks like "token1, token2"  then it seems this token2 is not going to be processed or I missed the fact that the use case will not happen.

Reply:
This code block handles the start and count params in the projection. 
For example: $*:(c),$start:2,$count:4

We are parsing for digits after we see the field name as "$start" or "$count" (see lines 189-190). 
So we move to the next token if we see "," or ")" (valid), parse the number if we see digit (valid) and anything else is invalid.

Reply:
This code block handles the start and count params in the projection. 
For example: $*:(c),$start:2,$count:4

We are parsing for digits after we see the field name as "$start" or "$count" (see lines 189-190). 
So we move to the next token if we see "," or ")" (valid), parse the number if we see digit (valid) and anything else is invalid.